codebase/analytics-gemini/client.go

```diff
@@ -0,0 +1,207 @@
+
+		codebaseSearchURL:    codebaseSearchURL,
+		codebaseSearchClient: codebaseSearchClient,
+	}
+}
+
+func (sr *ScanRepoClient) setHeaders(ctx context.Context, req *http.Request) {
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	token := goajwt.ContextJWT(ctx)
+	if token != nil {
+		req.Header.Set("Authorization", "Bearer "+goajwt.ContextJWT(ctx).Raw)
+	} else {
+		log.Warn(ctx, map[string]interface{}{
+			"err": "context has no JWT token",
+		}, "")
+	}
+
+	req.Header.Set(middleware.RequestIDHeader, middleware.ContextRequestID(ctx))
+}
+
+func logErrors(ctx context.Context, d *request, err error, bespokeError string) {
+	log.Error(ctx, map[string]interface{}{
+		"repoURL": d.gitURL,
+		"emailID": d.emailIDs,
+		"err":     err,
+	}, bespokeError)
+}
+
+func (sr *ScanRepoClient) callGemini(
+	ctx context.Context,
+	d *request,
+	path string,
+	finalResponse string,
+) error {
+	dBytes, err := json.Marshal(d)
+	if err != nil {
+		logErrors(ctx, d, err, "failed to marshal object into json")
+		return errs.WithStack(err)
+	}
+	data := bytes.NewReader(dBytes)
+
+	geminiURL, err := url.Parse(sr.geminiURL)
+	if err != nil {
+		return errs.WithStack(err)
+	}
+	geminiURL.Path = path
+
+	req, err := http.NewRequest("POST", geminiURL.String(), data)
+	if err != nil {
+		logErrors(ctx, d, err, "failed to create request object")
+		return errs.WithStack(err)
+	}
+	sr.setHeaders(ctx, req)
+
+	if log.IsDebug() {
+		b, _ := httputil.DumpRequest(req, true)
+		log.Debug(ctx, map[string]interface{}{
+			"request": string(b),
+		}, "request object")
+	}
+
+	resp, err := sr.geminiClient.Do(req)
+	if err != nil {
+		logErrors(ctx, d, err, "failed to talk to analytics gemini service")
+		return errs.WithStack(err)
+	}
+	defer rest.CloseResponse(resp)
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logErrors(ctx, d, err, "")
+		return errs.WithStack(err)
+	}
+
+	r := response{}
+	if err := json.Unmarshal(body, &r); err != nil {
+		logErrors(ctx, d, err, "failed to unmarshal the response")
+		return errs.WithStack(err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		switch resp.StatusCode {
+		case http.StatusUnauthorized:
+			return fmt.Errorf("unauthorized access: %v", r.Error)
+		default:
+			return fmt.Errorf("unknown error, got response: %s", string(body))
+		}
+	}
+
+	if r.Summary != finalResponse {
+		return fmt.Errorf("got unknown response: %s", string(body))
+	}
+
+	return nil
+}
+
+func (sr *ScanRepoClient) Register(ctx context.Context, d *request) error {
+	return sr.callGemini(ctx, d, "/api/v1/user-repo/scan", "Repository scan initiated")
+}
+
+func (sr *ScanRepoClient) DeRegister(ctx context.Context, d *request) error {
+	// first list all the records with this repoURL in the database
+	codebases, err := sr.listCodebases(ctx, d)
+	if err != nil {
+		return err
+	}
+
+	// if there are codebases with same url then don't stop the scan
+	if len(codebases.Data) > 0 {
```

Comment:
If there are Codebases that has scan active. The codebase can live in many spaces while only 1 has it active.

Reply:
what do you mean here?

Reply:
if you mean i am checking space specific codebases, so no i am querying for the codebase url and it lists all the codebases from all the space so we are good on this check i think!

Reply:
A Codebase URL is not unique across all Spaces, which means a Codebase Search can return multiple unique Codebases for the same URL across multiple spaces. Analytics only deal with 'Codebase URL'. That imply that even tho 'this' codebase is 'deactivated', you'll need to check if the other Codebases in the other Spaces are also 'deactived'. If one is still Active, then don't disable the scan.